sdk/core/azure-core-amqp/inc/azure/core/amqp/models/amqp_value.hpp

```diff
@@ -0,0 +1,928 @@
+     *
+     * @param value value to be set.
+     *
+     */
+    AmqpValue(std::int8_t value);
+    AmqpValue(char value);
+
+    /** @brief Construct an AMQP short value, a 16 bit signed integer.
+     *
+     * Defined in [AMQP Core Types
+     * section 1.6.8](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-short).
+     *
+     * @param value value to be set.
+     *
+     */
+    AmqpValue(std::int16_t value);
+
+    /** @brief Construct an AMQP int value, a 32 bit signed integer.
+     *
+     * Defined in [AMQP Core Types
+     * section 1.6.9](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-int).
+     *
+     * @param value value to be set.
+     *
+     */
+    AmqpValue(std::int32_t value);
+
+    /** @brief Construct an AMQP long value, a 64 bit signed integer.
+     *
+     * Defined in [AMQP Core Types
+     * section 1.6.10](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-long).
+     *
+     * @param value value to be set.
+     *
+     */
+    AmqpValue(std::int64_t value);
+
+    /** @brief Construct an AMQP float value, an IEEE 754-2008 value.
+     *
+     * Defined in [AMQP Core Types
+     * section 1.6.11](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-float).
+     *
+     * @param value value to be set.
+     *
+     */
+    AmqpValue(float value);
+
+    /** @brief Construct an AMQP double value, an IEEE 754-2008 value.
+     *
+     * Defined in [AMQP Core Types
+     * section 1.6.12](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-double).
+     *
+     * @param value value to be set.
+     *
+     */
+    AmqpValue(double value);
+
+    /** @brief Construct an AMQP string value, a UTF-8 encoded sequence of characters.
+     *
+     * Defined in [AMQP Core Types
+     * section 1.6.20](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-string).
+     *
+     * @param value to be set.
+     *
+     */
+    explicit AmqpValue(std::string const& value);
+
+    /** @brief Construct an AMQP string value, a UTF-8 encoded sequence of characters.
+     *
+     * Defined in [AMQP Core Types
+     * section 1.6.20](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-string).
+     *
+     * @param value to be set.
+     *
+     * @remarks This is a convenience constructor to allow callers to specify:
+     * ```cpp
+     * AmqpValue myValue("This is some text");
+     * ```
+     *
+     */
+    AmqpValue(const char* value);
+
+    /** @brief Construct an AMQP Char value, a UTF-32BE encoded Unicode character.
+     *
+     * Defined in [AMQP Core Types
+     * section 1.6.16](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-char).
+     *
+     * @param value UTF-32 encoded unicode value to be set.
+     *
+     */
+    AmqpValue(char32_t value);
```

Comment:
Will we actually need such APIs, in practice?

I get we seem to be wrapping all of uamqp, but maybe we should be selective (based on what we actually need), to avoid the risk of inheriting potentially buggy behavior/API surface of things we don't need from a general purpose library, given character encoding can be tricky.

Reply:
I don't believe we can claim to have an AMQP library if it doesn't support as much of the AMQP protocol as possible.

Remember that all the public model types *will* be exposed to customers (Clemens insists that we have mechanism to allow customers to send and receive raw AMQP messages). Customers will be rather upset if they try to use types defined in AMQP but we don't allow them to do that because we decided "nah, nobody is going to need this".

Having said that: There *are* bugs in uAMQP that I'm currently ignoring w.r.t. the functionality being built (message footers can't be sent but can be received, Char cannot be serialized or deserialized, etc). And we're going to have to work around those issues if customers notice.

Reply:
I don't understand what endianness means in UTF-32:
http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-char

The doc comment says it is UTF-32BE, does that mean we have to figure it out based on reading the BOM.  What endianness is char32_t? It seems to be platform dependent from a cursory look on SO.